Documentation build fails due to new sphinx version

Traceback
```
Run sphinx-build -b html docs/src build/docs/html -j auto
Running Sphinx v8.2.0
loading translations [en]... done
WARNING: while setting up extension sphinxcontrib.apidoc: Failed to convert [[<class 'str'>]] to a frozenset
making output directory... done
myst v4.0.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions={'colon_fence', 'substitution'}, disable_syntax=[], all_links_external=False, links_external_new_tab=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, fence_as_directive=set(), number_code_blocks=[], title_to_header=False, heading_anchors=0, heading_slug_func=None, html_meta={}, footnote_sort=True, footnote_transition=True, words_per_minute=200, substitutions={FmuResultsSchema: ..., InplaceVolumesSchema: ...}, linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', enable_checkboxes=False, suppress_warnings=[], highlight_code_blocks=True)

Extension error (pydantic_autosummary)!

Versions
========

* Platform:         linux; (Linux-6.8.0-1021-azure-x86_64-with-glibc2.39)
* Python version:   3.[11](https://github.com/equinor/fmu-dataio/actions/runs/13411808579/job/37463515502#step:6:12).11 (CPython)
* Sphinx version:   8.2.0
* Docutils version: 0.21.2
* Jinja2 version:   3.1.5
* Pygments version: 2.19.1

Last Messages
=============

None.

Loaded Extensions
=================

None.

Traceback
=========

      File "/opt/hostedtoolcache/Python/3.11.11/x64/lib/python3.11/site-packages/sphinx/events.py", line 4[15](https://github.com/equinor/fmu-dataio/actions/runs/13411808579/job/37463515502#step:6:16), in emit
        raise ExtensionError(
    sphinx.errors.ExtensionError: Handler <function process_generate_options at 0x7fdbc6c8ef20> for event 'builder-inited' threw an exception (exception: BuildEnvironment.doctreedir may only be set to path-like objects)


The full traceback has been saved in:
/tmp/sphinx-err-yzs4_q65.log

To report this error to the developers, please open an issue at <https://github.com/sphinx-doc/sphinx/issues/>. Thanks!
Please also report this if it was a user error, so that a better error message can be provided next time.
Error: Process completed with exit code 2.
```

--- a/docs/ext/pydantic_autosummary/__init__.py
+++ b/docs/ext/pydantic_autosummary/__init__.py
@@ -11,7 +11,7 @@ autosummary directive
 The autosummary directive has the form::
 
     .. autosummary::
-       :nosignatures:
+       :signatures: none
        :toctree: generated/
 
        module.function_1
@@ -51,19 +51,23 @@ from __future__ import annotations
 import functools
 import inspect
 import operator
-import os
 import posixpath
 import re
 import sys
 from inspect import Parameter
 from os import path
+from pathlib import Path
 from types import ModuleType
 from typing import TYPE_CHECKING, Any, ClassVar, cast
 
 import sphinx
 from docutils import nodes
-from docutils.parsers.rst import directives
-from docutils.parsers.rst.states import RSTStateMachine, Struct, state_classes
+from docutils.parsers.rst import directives  # type: ignore
+from docutils.parsers.rst.states import (  # type: ignore
+    RSTStateMachine,
+    Struct,
+    state_classes,
+)
 from docutils.statemachine import StringList
 from sphinx import addnodes
 from sphinx.config import Config
@@ -86,6 +90,7 @@ from sphinx.util.docutils import (
 )
 from sphinx.util.inspect import getmro, signature_from_str
 from sphinx.util.matching import Matcher
+from sphinx.util.parsing import nested_parse_to_nodes
 
 if TYPE_CHECKING:
     from collections.abc import Sequence
@@ -155,10 +160,10 @@ class FakeApplication:
     verbosity = 0
 
     def __init__(self) -> None:
-        self.doctreedir = None
+        self.doctreedir = Path()
         self.events = None
         self.extensions: dict[str, Extension] = {}
-        self.srcdir = None
+        self.srcdir = Path()
         self.config = Config()
         self.project = Project("", {})
         self.registry = SphinxComponentRegistry()
@@ -183,6 +188,19 @@ def get_documenter(app: Sphinx, obj: Any, parent: Any) -> type[Documenter]:
     another Python object (e.g. a module or a class) to which *obj*
     belongs to.
     """
+    return _get_documenter(obj, parent, registry=app.registry)
+
+
+def _get_documenter(
+    obj: Any, parent: Any, *, registry: SphinxComponentRegistry
+) -> type[Documenter]:
+    """Get an autodoc.Documenter class suitable for documenting the given
+    object.
+
+    *obj* is the Python object to be documented, and *parent* is an
+    another Python object (e.g. a module or a class) to which *obj*
+    belongs to.
+    """
     from sphinx.ext.autodoc import DataDocumenter, ModuleDocumenter
 
     if inspect.ismodule(obj):
@@ -191,7 +209,7 @@ def get_documenter(app: Sphinx, obj: Any, parent: Any) -> type[Documenter]:
 
     # Construct a fake documenter for *parent*
     if parent is not None:
-        parent_doc_cls = get_documenter(app, parent, None)
+        parent_doc_cls = _get_documenter(parent, None, registry=registry)
     else:
         parent_doc_cls = ModuleDocumenter
 
@@ -203,7 +221,7 @@ def get_documenter(app: Sphinx, obj: Any, parent: Any) -> type[Documenter]:
     # Get the correct documenter class for *obj*
     classes = [
         cls
-        for cls in app.registry.documenters.values()
+        for cls in registry.documenters.values()
         if cls.can_document_member(obj, "", False, parent_doc)
     ]
     if classes:
@@ -228,9 +246,11 @@ class Autosummary(SphinxDirective):
     has_content = True
     option_spec: ClassVar[OptionSpec] = {
         "caption": directives.unchanged_required,
+        "class": directives.class_option,
         "toctree": directives.unchanged,
         "nosignatures": directives.flag,
         "recursive": directives.flag,
+        "signatures": directives.unchanged,
         "template": directives.unchanged,
     }
 
@@ -259,7 +279,7 @@ class Autosummary(SphinxDirective):
                 docname = posixpath.join(tree_prefix, real_name)
                 docname = posixpath.normpath(posixpath.join(dirname, docname))
                 if docname not in self.env.found_docs:
-                    if excluded(self.env.doc2path(docname, False)):
+                    if excluded(str(self.env.doc2path(docname, False))):
                         msg = __(
                             "autosummary references excluded document %r. Ignored."
                         )
@@ -313,14 +333,19 @@ class Autosummary(SphinxDirective):
                     raise ImportExceptionGroup(exc.args[0], errors) from None
 
     def create_documenter(
-        self, app: Sphinx, obj: Any, parent: Any, full_name: str
+        self,
+        obj: Any,
+        parent: Any,
+        full_name: str,
+        *,
+        registry: SphinxComponentRegistry,
     ) -> Documenter:
         """Get an autodoc.Documenter class suitable for documenting the given
         object.
 
-        Wraps get_documenter and is meant as a hook for extensions.
+        Wraps _get_documenter and is meant as a hook for extensions.
         """
-        doccls = get_documenter(app, obj, parent)
+        doccls = _get_documenter(obj, parent, registry=registry)
         return doccls(self.bridge, full_name)
 
     def get_items(self, names: list[str]) -> list[tuple[str, str, str, str]]:
@@ -330,6 +355,16 @@ class Autosummary(SphinxDirective):
         prefixes = get_import_prefixes_from_env(self.env)
 
         items: list[tuple[str, str, str, str]] = []
+
+        signatures_option = self.options.get("signatures")
+        if signatures_option is None:
+            signatures_option = "none" if "nosignatures" in self.options else "long"
+        if signatures_option not in {"none", "short", "long"}:
+            msg = (
+                "Invalid value for autosummary :signatures: option: "
+                f"{signatures_option!r}. Valid values are 'none', 'short', 'long'"
+            )
+            raise ValueError(msg)
 
         max_item_chars = 50
 
@@ -361,7 +396,12 @@ class Autosummary(SphinxDirective):
                 full_name = modname + "::" + full_name[len(modname) + 1 :]
             # NB. using full_name here is important, since Documenters
             #     handle module prefixes slightly differently
-            documenter = self.create_documenter(self.env.app, obj, parent, full_name)
+            documenter = self.create_documenter(
+                obj,
+                parent,
+                full_name,
+                registry=self.env._registry,
+            )
             if not documenter.parse_name():
                 logger.warning(
                     __("failed to parse name %s"),
@@ -394,6 +434,8 @@ class Autosummary(SphinxDirective):
 
             # -- Grab the signature
 
+            if signatures_option == "none":
+                sig = None
             try:
                 sig = documenter.format_signature(show_annotation=False)
             except TypeError:
@@ -402,6 +444,9 @@ class Autosummary(SphinxDirective):
 
             if not sig:
                 sig = ""
+            elif signatures_option == "short":
+                if sig != "()":
+                    sig = "(…)"
             else:
                 max_chars = max(10, max_item_chars - len(display_name))
                 sig = mangle_signature(sig, max_chars=max_chars)
@@ -418,7 +463,7 @@ class Autosummary(SphinxDirective):
 
         return items
 
-    def get_table(self, items: list[tuple[str, str, str, str]]) -> list[Node]:
+    def get_table(self, items: list[tuple[str, str | None, str, str]]) -> list[Node]:
         """Generate a proper list of table nodes for autosummary:: directive.
 
         *items* is a list produced by :meth:`get_items`.
@@ -427,7 +472,9 @@ class Autosummary(SphinxDirective):
         table_spec["spec"] = r"\X{1}{2}\X{1}{2}"
 
         table = autosummary_table("")
-        real_table = nodes.table("", classes=["autosummary longtable"])
+        real_table = nodes.table(
+            "", classes=["autosummary", "longtable", *self.options.get("class", ())]
+        )
         table.append(real_table)
         group = nodes.tgroup("", cols=2)
         real_table.append(group)
@@ -440,25 +487,26 @@ class Autosummary(SphinxDirective):
             row = nodes.row("")
             source, line = self.state_machine.get_source_and_line()
             for text in column_texts:
-                node = nodes.paragraph("")
-                vl = StringList()
-                vl.append(text, "%s:%d:<autosummary>" % (source, line))
+                vl = StringList([text], f"{source}:{line}:<autosummary>")
                 with switch_source_input(self.state, vl):
-                    self.state.nested_parse(vl, 0, node)
-                    try:
-                        if isinstance(node[0], nodes.paragraph):
-                            node = node[0]
-                    except IndexError:
-                        pass
+                    col_nodes = nested_parse_to_nodes(
+                        self.state, vl, allow_section_headings=False
+                    )
+                    if col_nodes and isinstance(col_nodes[0], nodes.paragraph):
+                        node = col_nodes[0]
+                    else:
+                        node = nodes.paragraph("")
                     row.append(nodes.entry("", node))
+                node = nodes.paragraph("")
             body.append(row)
 
         for name, sig, summary, real_name in items:
             qualifier = "obj"
-            if "nosignatures" not in self.options:
-                col1 = f":py:{qualifier}:`{name} <{real_name}>`\\ {rst.escape(sig)}"
-            else:
+            if sig is None:
                 col1 = f":py:{qualifier}:`{name} <{real_name}>`"
+            else:
+                col1 = f":py:{qualifier}:`{name} <{real_name}>`\\ {rst.escape(sig)}"
+
             col2 = summary
             append_row(col1, col2)
 
@@ -543,13 +591,13 @@ def mangle_signature(sig: str, max_chars: int = 30) -> str:
     sig = limited_join(", ", args, max_chars=max_chars - 2)
     if opts:
         if not sig:
-            sig = "[%s]" % limited_join(", ", opts, max_chars=max_chars - 4)
+            sig = "[%s]" % limited_join(", ", opts, max_chars=max_chars - 4)  # noqa
         elif len(sig) < max_chars - 4 - 2 - 3:
-            sig += "[, %s]" % limited_join(
+            sig += "[, %s]" % limited_join(  # noqa
                 ", ", opts, max_chars=max_chars - len(sig) - 4 - 2
             )
 
-    return "(%s)" % sig
+    return "(%s)" % sig  # noqa
 
 
 def extract_summary(doc: list[str], document: Any) -> str:
@@ -664,7 +712,7 @@ def get_import_prefixes_from_env(env: BuildEnvironment) -> list[str | None]:
     currclass = env.ref_context.get("py:class")
     if currclass:
         if currmodule:
-            prefixes.insert(0, currmodule + "." + currclass)
+            prefixes.insert(0, f"{currmodule}.{currclass}")
         else:
             prefixes.insert(0, currclass)
 
@@ -681,6 +729,20 @@ def import_by_name(
     tried = []
     errors: list[ImportExceptionGroup] = []
     for prefix in prefixes:
+        if prefix is not None and name.startswith(f"{prefix}."):
+            # Catch and avoid module cycles (e.g., sphinx.ext.sphinx.ext...)
+            msg = __(
+                "Summarised items should not include the current module. "
+                "Replace %r with %r."
+            )
+            logger.warning(
+                msg,
+                name,
+                name.removeprefix(f"{prefix}."),
+                type="autosummary",
+                subtype="import_cycle",
+            )
+            continue
         try:
             prefixed_name = f"{prefix}.{name}" if prefix else name
             obj, parent, modname = _import_by_name(
@@ -696,7 +758,7 @@ def import_by_name(
     exceptions: list[BaseException] = functools.reduce(
         operator.iadd, (e.exceptions for e in errors), []
     )
-    raise ImportExceptionGroup("no module named %s" % " or ".join(tried), exceptions)
+    raise ImportExceptionGroup("no module named %s" % " or ".join(tried), exceptions)  # noqa
 
 
 def _import_by_name(name: str, grouped_exception: bool = True) -> tuple[Any, Any, str]:
@@ -770,7 +832,7 @@ def import_ivar_by_name(
             found_attrs |= {attr for (qualname, attr) in analyzer.attr_docs}
             found_attrs |= {attr for (qualname, attr) in analyzer.annotations}
             if attr in found_attrs:
-                return real_name + "." + attr, INSTANCEATTR, obj, modname
+                return f"{real_name}.{attr}", INSTANCEATTR, obj, modname
     except (ImportError, ValueError, PycodeError) as exc:
         raise ImportError from exc
     except ImportExceptionGroup:
@@ -790,7 +852,7 @@ class AutoLink(SphinxRole):
     """
 
     def run(self) -> tuple[list[Node], list[system_message]]:
-        pyobj_role = self.env.get_domain("py").role("obj")
+        pyobj_role = self.env.domains.python_domain.role("obj")
         assert pyobj_role is not None
         objects, errors = pyobj_role(
             "obj",
@@ -805,13 +867,20 @@ class AutoLink(SphinxRole):
             return objects, errors
 
         assert len(objects) == 1
-        pending_xref = cast(addnodes.pending_xref, objects[0])
+        pending_xref = cast("addnodes.pending_xref", objects[0])
         try:
             # try to import object by name
             prefixes = get_import_prefixes_from_env(self.env)
+            name = pending_xref["reftarget"]
+            prefixes = [
+                prefix
+                for prefix in prefixes
+                if prefix is None
+                or not (name.startswith(f"{prefix}.") or name == prefix)
+            ]
             import_by_name(pending_xref["reftarget"], prefixes)
         except ImportExceptionGroup:
-            literal = cast(nodes.literal, pending_xref[0])
+            literal = cast("nodes.literal", pending_xref[0])
             objects[0] = nodes.emphasis(
                 self.rawtext, literal.astext(), classes=literal["classes"]
             )
@@ -838,11 +907,11 @@ def process_generate_options(app: Sphinx) -> None:
     genfiles = app.config.autosummary_generate
 
     if genfiles is True:
-        env = app.builder.env
+        env = app.env
         genfiles = [
-            env.doc2path(x, base=False)
+            str(env.doc2path(x, base=False))
             for x in env.found_docs
-            if os.path.isfile(env.doc2path(x))
+            if env.doc2path(x).is_file()
         ]
     elif genfiles is False:
         pass
@@ -910,14 +979,30 @@ def setup(app: Sphinx) -> ExtensionMetadata:
     app.add_directive("autosummary", Autosummary)
     app.add_role("autolink", AutoLink())
     app.connect("builder-inited", process_generate_options)
-    app.add_config_value("autosummary_context", {}, "env")
-    app.add_config_value("autosummary_filename_map", {}, "html")
-    app.add_config_value("autosummary_generate", True, "env", {bool, list})
-    app.add_config_value("autosummary_generate_overwrite", True, "")
+    app.add_config_value("autosummary_context", {}, "env", types=frozenset({dict}))
     app.add_config_value(
-        "autosummary_mock_imports", lambda config: config.autodoc_mock_imports, "env"
+        "autosummary_filename_map", {}, "html", types=frozenset({dict})
     )
-    app.add_config_value("autosummary_imported_members", [], "", bool)
-    app.add_config_value("autosummary_ignore_module_all", True, "env", bool)
+    app.add_config_value(
+        "autosummary_generate", True, "env", types=frozenset({bool, list})
+    )
+    app.add_config_value(
+        "autosummary_generate_overwrite", True, "", types=frozenset({bool})
+    )
+    app.add_config_value(
+        "autosummary_mock_imports",
+        lambda config: config.autodoc_mock_imports,
+        "env",
+        types=frozenset({list, tuple}),
+    )
+    app.add_config_value(
+        "autosummary_imported_members", False, "", types=frozenset({bool})
+    )
+    app.add_config_value(
+        "autosummary_ignore_module_all", True, "env", types=frozenset({bool})
+    )
 
-    return {"version": sphinx.__display_version__, "parallel_read_safe": True}
+    return {
+        "version": sphinx.__display_version__,
+        "parallel_read_safe": True,
+    }

--- a/docs/ext/pydantic_autosummary/generate.py
+++ b/docs/ext/pydantic_autosummary/generate.py
@@ -23,9 +23,8 @@ import pkgutil
 import pydoc
 import re
 import sys
-from os import path
 from pathlib import Path
-from typing import TYPE_CHECKING, Any, NamedTuple
+from typing import TYPE_CHECKING, NamedTuple
 
 import sphinx.locale
 from jinja2 import TemplateNotFound
@@ -33,11 +32,13 @@ from jinja2.sandbox import SandboxedEnvironment
 from sphinx import __display_version__
 from sphinx.builders import Builder
 from sphinx.config import Config
+from sphinx.errors import PycodeError
 from sphinx.ext.autodoc.importer import import_module
 from sphinx.locale import __
-from sphinx.pycode import ModuleAnalyzer, PycodeError
+from sphinx.pycode import ModuleAnalyzer
 from sphinx.registry import SphinxComponentRegistry
 from sphinx.util import logging, rst
+from sphinx.util._pathlib import _StrPath
 from sphinx.util.inspect import getall, safe_getattr
 from sphinx.util.osutil import ensuredir
 from sphinx.util.template import SphinxTemplateLoader
@@ -45,7 +46,7 @@ from sphinx.util.template import SphinxTemplateLoader
 # ----------- pydantic_autosummary change
 from . import (
     ImportExceptionGroup,
-    get_documenter,
+    _get_documenter,
     import_by_name,
     import_ivar_by_name,
 )
@@ -56,11 +57,18 @@ from .pydantic import set_pydantic_model_fields
 if TYPE_CHECKING:
     from collections.abc import Sequence, Set
     from gettext import NullTranslations
+    from typing import Any
 
     from sphinx.application import Sphinx
+    from sphinx.events import EventManager
     from sphinx.ext.autodoc import Documenter
 
 logger = logging.getLogger(__name__)
+
+
+class _DummyEvents:
+    def emit_firstresult(self, *args: Any) -> None:
+        pass
 
 
 class DummyApplication:
@@ -68,13 +76,14 @@ class DummyApplication:
 
     def __init__(self, translator: NullTranslations) -> None:
         self.config = Config()
+        self.events = _DummyEvents()
         self.registry = SphinxComponentRegistry()
         self.messagelog: list[str] = []
-        self.srcdir = "/"
+        self.srcdir = _StrPath("/")
         self.translator = translator
         self.verbosity = 0
         self._warncount = 0
-        self.warningiserror = False
+        self._exception_on_warning = False
 
         self.config.add("autosummary_context", {}, "env", ())
         self.config.add("autosummary_filename_map", {}, "env", ())
@@ -91,7 +100,7 @@ class AutosummaryEntry(NamedTuple):
     recursive: bool
 
 
-def setup_documenters(app: Any) -> None:
+def setup_documenters(app: Sphinx) -> None:
     from sphinx.ext.autodoc import (
         AttributeDocumenter,
         ClassDocumenter,
@@ -132,10 +141,10 @@ class AutosummaryRenderer:
     def __init__(self, app: Sphinx) -> None:
         if isinstance(app, Builder):
             msg = "Expected a Sphinx application object!"
-            raise ValueError(msg)
+            raise TypeError(msg)
 
         # ----------- pydantic_autosummary change
-        system_templates_path = [os.path.join(os.path.dirname(__file__), "templates")]
+        system_templates_path = [Path(__file__).parent / "templates"]
         # ----------/ pydantic_autosummary change
         loader = SphinxTemplateLoader(
             app.srcdir, app.config.templates_path, system_templates_path
@@ -159,7 +168,7 @@ class AutosummaryRenderer:
         except TemplateNotFound:
             try:
                 # objtype is given as template_name
-                template = self.env.get_template("autosummary/%s.rst" % template_name)
+                template = self.env.get_template("autosummary/%s.rst" % template_name)  # noqa
             except TemplateNotFound:
                 # fallback to base.rst
                 template = self.env.get_template("autosummary/base.rst")
@@ -200,16 +209,25 @@ def _split_full_qualified_name(name: str) -> tuple[str | None, str]:
 
 
 class ModuleScanner:
-    def __init__(self, app: Any, obj: Any) -> None:
-        self.app = app
+    def __init__(
+        self,
+        obj: Any,
+        *,
+        config: Config,
+        events: EventManager,
+        registry: SphinxComponentRegistry,
+    ) -> None:
+        self.config = config
+        self.events = events
+        self.registry = registry
         self.object = obj
 
     def get_object_type(self, name: str, value: Any) -> str:
-        return get_documenter(self.app, value, self.object).objtype
+        return _get_documenter(value, self.object, registry=self.registry).objtype
 
     def is_skipped(self, name: str, value: Any, objtype: str) -> bool:
         try:
-            return self.app.emit_firstresult(
+            return self.events.emit_firstresult(
                 "autodoc-skip-member", objtype, name, value, False, {}
             )
         except Exception as exc:
@@ -234,7 +252,7 @@ class ModuleScanner:
         except PycodeError:
             attr_docs = {}
 
-        for name in members_of(self.object, self.app.config):
+        for name in members_of(self.object, config=self.config):
             try:
                 value = safe_getattr(self.object, name)
             except AttributeError:
@@ -256,7 +274,7 @@ class ModuleScanner:
             except AttributeError:
                 imported = False
 
-            respect_module_all = not self.app.config.autosummary_ignore_module_all
+            respect_module_all = not self.config.autosummary_ignore_module_all
             if (
                 # list all members up
                 imported_members
@@ -270,12 +288,18 @@ class ModuleScanner:
         return members
 
 
-def members_of(obj: Any, conf: Config) -> Sequence[str]:
+def members_of(obj: Any, *, config: Config) -> Sequence[str]:
     """Get the members of ``obj``, possibly ignoring the ``__all__`` module attribute
 
-    Follows the ``conf.autosummary_ignore_module_all`` setting.
+    Follows the ``config.autosummary_ignore_module_all`` setting.
     """
-    return dir(obj) if conf.autosummary_ignore_module_all else (getall(obj) or dir(obj))
+    if config.autosummary_ignore_module_all:
+        return dir(obj)
+    if (obj___all__ := getall(obj)) is not None:
+        # return __all__, even if empty.
+        return obj___all__
+    # if __all__ is not set, return dir(obj)
+    return dir(obj)
 
 
 def generate_autosummary_content(
@@ -285,44 +309,77 @@ def generate_autosummary_content(
     template: AutosummaryRenderer,
     template_name: str,
     imported_members: bool,
-    app: Any,
     recursive: bool,
     context: dict[str, Any],
     modname: str | None = None,
     qualname: str | None = None,
+    *,
+    config: Config,
+    events: EventManager,
+    registry: SphinxComponentRegistry,
 ) -> str:
-    doc = get_documenter(app, obj, parent)
+    doc = _get_documenter(obj, parent, registry=registry)
 
     ns: dict[str, Any] = {}
     ns.update(context)
 
     if doc.objtype == "module":
-        scanner = ModuleScanner(app, obj)
+        scanner = ModuleScanner(obj, config=config, events=events, registry=registry)
         ns["members"] = scanner.scan(imported_members)
 
-        respect_module_all = not app.config.autosummary_ignore_module_all
+        respect_module_all = not config.autosummary_ignore_module_all
         imported_members = imported_members or (
             "__all__" in dir(obj) and respect_module_all
         )
 
         ns["functions"], ns["all_functions"] = _get_members(
-            doc, app, obj, {"function"}, imported=imported_members
+            doc,
+            obj,
+            {"function"},
+            config=config,
+            events=events,
+            registry=registry,
+            imported=imported_members,
         )
         ns["classes"], ns["all_classes"] = _get_members(
-            doc, app, obj, {"class"}, imported=imported_members
+            doc,
+            obj,
+            {"class"},
+            config=config,
+            events=events,
+            registry=registry,
+            imported=imported_members,
         )
 
         # ----------- pydantic_autosummary change
         ns["pydantic_models"], ns["all_pydantic_models"] = _get_members(
-            doc, app, obj, {"pydantic_model"}, imported=imported_members
+            doc,
+            obj,
+            {"pydantic_model"},
+            config=config,
+            events=events,
+            registry=registry,
+            imported=imported_members,
         )
         ns["pydantic_settings"], ns["all_pydantic_settings"] = _get_members(
-            doc, app, obj, {"pydantic_settings"}, imported=imported_members
+            doc,
+            obj,
+            {"pydantic_settings"},
+            config=config,
+            events=events,
+            registry=registry,
+            imported=imported_members,
         )
         # ----------/ pydantic_autosummary change
 
         ns["exceptions"], ns["all_exceptions"] = _get_members(
-            doc, app, obj, {"exception"}, imported=imported_members
+            doc,
+            obj,
+            {"exception"},
+            config=config,
+            events=events,
+            registry=registry,
+            imported=imported_members,
         )
         ns["attributes"], ns["all_attributes"] = _get_module_attrs(name, ns["members"])
 
@@ -345,15 +402,15 @@ def generate_autosummary_content(
             # Otherwise, use get_modules method normally
             if respect_module_all and "__all__" in dir(obj):
                 imported_modules, all_imported_modules = _get_members(
-                    doc, app, obj, {"module"}, imported=True
+                    doc,
+                    obj,
+                    {"module"},
+                    config=config,
+                    events=events,
+                    registry=registry,
+                    imported=True,
                 )
                 skip += all_imported_modules
-                imported_modules = [
-                    name + "." + modname for modname in imported_modules
-                ]
-                all_imported_modules = [
-                    name + "." + modname for modname in all_imported_modules
-                ]
                 public_members = getall(obj)
             else:
                 imported_modules, all_imported_modules = [], []
@@ -368,16 +425,27 @@ def generate_autosummary_content(
         ns["members"] = dir(obj)
         ns["inherited_members"] = set(dir(obj)) - set(obj.__dict__.keys())
         ns["methods"], ns["all_methods"] = _get_members(
-            doc, app, obj, {"method"}, include_public={"__init__"}
+            doc,
+            obj,
+            {"method"},
+            config=config,
+            events=events,
+            registry=registry,
+            include_public={"__init__"},
         )
         ns["attributes"], ns["all_attributes"] = _get_members(
-            doc, app, obj, {"attribute", "property"}
+            doc,
+            obj,
+            {"attribute", "property"},
+            config=config,
+            events=events,
+            registry=registry,
         )
 
     if modname is None or qualname is None:
         modname, qualname = _split_full_qualified_name(name)
 
-    if doc.objtype in ("method", "attribute", "property"):
+    if doc.objtype in {"method", "attribute", "property"}:
         ns["class"] = qualname.rsplit(".", 1)[0]
 
     # ----------- pydantic_autosummary change
@@ -398,9 +466,9 @@ def generate_autosummary_content(
     return template.render(template_name or doc.objtype, ns)
 
 
-def _skip_member(app: Sphinx, obj: Any, name: str, objtype: str) -> bool:
+def _skip_member(obj: Any, name: str, objtype: str, *, events: EventManager) -> bool:
     try:
-        return app.emit_firstresult(
+        return events.emit_firstresult(
             "autodoc-skip-member", objtype, name, obj, False, {}
         )
     except Exception as exc:
@@ -423,9 +491,9 @@ def _get_class_members(obj: Any) -> dict[str, Any]:
     return {name: member.object for name, member in members.items()}
 
 
-def _get_module_members(app: Sphinx, obj: Any) -> dict[str, Any]:
+def _get_module_members(obj: Any, *, config: Config) -> dict[str, Any]:
     members = {}
-    for name in members_of(obj, app.config):
+    for name in members_of(obj, config=config):
         try:
             members[name] = safe_getattr(obj, name)
         except AttributeError:
@@ -433,9 +501,11 @@ def _get_module_members(app: Sphinx, obj: Any) -> dict[str, Any]:
     return members
 
 
-def _get_all_members(doc: type[Documenter], app: Sphinx, obj: Any) -> dict[str, Any]:
+def _get_all_members(
+    doc: type[Documenter], obj: Any, *, config: Config
+) -> dict[str, Any]:
     if doc.objtype == "module":
-        return _get_module_members(app, obj)
+        return _get_module_members(obj, config=config)
     if doc.objtype == "class":
         return _get_class_members(obj)
     return {}
@@ -443,23 +513,25 @@ def _get_all_members(doc: type[Documenter], app: Sphinx, obj: Any) -> dict[str, 
 
 def _get_members(
     doc: type[Documenter],
-    app: Sphinx,
     obj: Any,
     types: set[str],
     *,
+    config: Config,
+    events: EventManager,
+    registry: SphinxComponentRegistry,
     include_public: Set[str] = frozenset(),
     imported: bool = True,
 ) -> tuple[list[str], list[str]]:
     items: list[str] = []
     public: list[str] = []
 
-    all_members = _get_all_members(doc, app, obj)
+    all_members = _get_all_members(doc, obj, config=config)
     for name, value in all_members.items():
-        documenter = get_documenter(app, value, obj)
+        documenter = _get_documenter(value, obj, registry=registry)
         if documenter.objtype in types and (
             imported or getattr(value, "__module__", None) == obj.__name__
         ):
-            skipped = _skip_member(app, value, name, documenter.objtype)
+            skipped = _skip_member(value, name, documenter.objtype, events=events)
             if skipped is True:
                 pass
             elif skipped is False:
@@ -481,7 +553,7 @@ def _get_module_attrs(name: str, members: Any) -> tuple[list[str], list[str]]:
         analyzer = ModuleAnalyzer.for_module(name)
         attr_docs = analyzer.find_attr_docs()
         for namespace, attr_name in attr_docs:
-            if namespace == "" and attr_name in members:
+            if not namespace and attr_name in members:
                 attrs.append(attr_name)
                 if not attr_name.startswith("_"):
                     public.append(attr_name)
@@ -503,7 +575,7 @@ def _get_modules(
         if modname in skip:
             # module was overwritten in __init__.py, so not accessible
             continue
-        fullname = name + "." + modname
+        fullname = f"{name}.{modname}"
         try:
             module = import_module(fullname)
         except ImportError:
@@ -527,7 +599,7 @@ def generate_autosummary_docs(
     suffix: str = ".rst",
     base_path: str | os.PathLike[str] | None = None,
     imported_members: bool = False,
-    app: Any = None,
+    app: Sphinx | None = None,
     overwrite: bool = True,
     encoding: str = "utf-8",
 ) -> list[Path]:
@@ -542,23 +614,30 @@ def generate_autosummary_docs(
         showed_sources = showed_sources[:10] + ["..."] + showed_sources[-10:]
 
     # ----------- pydantic_autosummary change
+    logger.info(
+        __("[pydantic_autosummary] generating autosummary for: %s"),
+        ", ".join(showed_sources),
+    )
     if output_dir:
         logger.info(__("[pydantic_autosummary] writing to %s") % output_dir)
     # ----------/ pydantic_autosummary change
 
     if base_path is not None:
-        sources = [os.path.join(base_path, filename) for filename in sources]
+        base_path = Path(base_path)
+        source_paths = [base_path / filename for filename in sources]
+    else:
+        source_paths = list(map(Path, sources))
 
     template = AutosummaryRenderer(app)
 
     # read
-    items = find_autosummary_in_files(sources)
+    items = find_autosummary_in_files(source_paths)
 
     # keep track of new files
     new_files: list[Path] = []
     all_files: list[Path] = []
 
-    filename_map = app.config.autosummary_filename_map if app else {}
+    filename_map = app.config.autosummary_filename_map
 
     # write
     for entry in sorted(set(items), key=str):
@@ -567,7 +646,7 @@ def generate_autosummary_docs(
             # a :toctree: option
             continue
 
-        path = output_dir or os.path.abspath(entry.path)
+        path = output_dir or Path(entry.path).resolve()
         ensuredir(path)
 
         try:
@@ -606,11 +685,13 @@ def generate_autosummary_docs(
             template,
             entry.template,
             imported_members,
-            app,
             entry.recursive,
             context,
             modname,
             qualname,
+            config=app.config,
+            events=app.events,
+            registry=app.registry,
         )
 
         file_path = Path(path, filename_map.get(name, name) + suffix)
@@ -650,7 +731,9 @@ def generate_autosummary_docs(
 # -- Finding documented entries in files ---------------------------------------
 
 
-def find_autosummary_in_files(filenames: list[str]) -> list[AutosummaryEntry]:
+def find_autosummary_in_files(
+    filenames: Sequence[str | os.PathLike[str]],
+) -> list[AutosummaryEntry]:
     """Find out what items are documented in source/*.rst.
 
     See `find_autosummary_in_lines`.
@@ -665,7 +748,7 @@ def find_autosummary_in_files(filenames: list[str]) -> list[AutosummaryEntry]:
 
 def find_autosummary_in_docstring(
     name: str,
-    filename: str | None = None,
+    filename: str | os.PathLike[str] | None = None,
 ) -> list[AutosummaryEntry]:
     """Find out what items are documented in the given object's docstring.
 
@@ -692,7 +775,7 @@ def find_autosummary_in_docstring(
 def find_autosummary_in_lines(
     lines: list[str],
     module: str | None = None,
-    filename: str | None = None,
+    filename: str | os.PathLike[str] | None = None,
 ) -> list[AutosummaryEntry]:
     """Find out what items appear in autosummary:: directives in the
     given lines.
@@ -732,7 +815,7 @@ def find_autosummary_in_lines(
             if m:
                 toctree = m.group(1)
                 if filename:
-                    toctree = os.path.join(os.path.dirname(filename), toctree)
+                    toctree = str(Path(filename).parent / toctree)
                 continue
 
             m = template_arg_re.match(line)
@@ -745,11 +828,9 @@ def find_autosummary_in_lines(
 
             m = autosummary_item_re.match(line)
             if m:
-                name = m.group(1).strip()
-                if name.startswith("~"):
-                    name = name[1:]
+                name = m.group(1).strip().removeprefix("~")
                 if current_module and not name.startswith(current_module + "."):
-                    name = f"{name}"
+                    name = str(name)
                 documented.append(AutosummaryEntry(name, toctree, template, recursive))
                 continue
 
@@ -806,7 +887,7 @@ The format of the autosummary directive is documented in the
         "--version",
         action="version",
         dest="show_version",
-        version="%%(prog)s %s" % __display_version__,
+        version="%%(prog)s %s" % __display_version__,  # noqa
     )
 
     parser.add_argument(
@@ -826,7 +907,7 @@ The format of the autosummary directive is documented in the
         action="store",
         dest="suffix",
         default="rst",
-        help=__("default suffix for files (default: " "%(default)s)"),
+        help=__("default suffix for files (default: %(default)s)"),
     )
     parser.add_argument(
         "-t",
@@ -834,7 +915,7 @@ The format of the autosummary directive is documented in the
         action="store",
         dest="templates",
         default=None,
-        help=__("custom template directory (default: " "%(default)s)"),
+        help=__("custom template directory (default: %(default)s)"),
     )
     parser.add_argument(
         "-i",
@@ -842,7 +923,7 @@ The format of the autosummary directive is documented in the
         action="store_true",
         dest="imported_members",
         default=False,
-        help=__("document imported members (default: " "%(default)s)"),
+        help=__("document imported members (default: %(default)s)"),
     )
     parser.add_argument(
         "-a",
@@ -878,7 +959,7 @@ def main(argv: Sequence[str] = (), /) -> None:
     args = get_parser().parse_args(argv or sys.argv[1:])
 
     if args.templates:
-        app.config.templates_path.append(path.abspath(args.templates))
+        app.config.templates_path.append(str(Path(args.templates).resolve()))
     app.config.autosummary_ignore_module_all = not args.respect_module_all
 
     written_files = generate_autosummary_docs(

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -66,13 +66,13 @@ dev = [
     "types-PyYAML",
 ]
 docs = [
-    "Sphinx==8.1.3",
+    "Sphinx",
     "autoapi",
     "autodoc-pydantic>=2.0.0",
     "furo",
     "myst-parser",
     "pydocstyle",
-    "sphinx-autodoc-typehints<1.23",
+    "sphinx-autodoc-typehints",
     "sphinx-copybutton",
     "sphinx-togglebutton",
     "sphinxcontrib-apidoc",

